### PR TITLE
Feat/result streaming - Fix for issue #746 in toolsservice and issue Microsoft/azuredatastudio#3348

### PR DESCRIPTION
**Pull request by ranasaria**
This pull request fixes the timing issues that was leading to flaky test behavior and duplicate ResulSetAvailable notification firing.

Issues fixed:
https://github.com/Microsoft/azuredatastudio/issues/3348
https://github.com/Microsoft/sqltoolsservice/issues/746

### CODE REVIEW

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs [hunk 1]
```diff
@@ -613,45 +617,70 @@ protected virtual void Dispose(bool disposing)
         /// Sends the ResultsUpdated message if the number of rows has changed since last send.
         /// </summary>
         /// <param name="stateInfo"></param>
-        private async void SendResultAvailableOrUpdated (object stateInfo = null)
+        private void SendResultAvailableOrUpdated (object stateInfo = null)
```
Comment (MattIrv) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs]: looks like this doesn't need the `stateInfo` parameter anymore

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/QueryExecution/Common.cs [hunk 2]
```diff
@@ -55,18 +68,53 @@ public static class Common
...
+        public static IEnumerable<TestResultSet> TestResultSetsEnumeration
+        {
+            get
+            {
+                yield return StandardTestResultSet;
+                yield return  HundredRowTestResultSet;
+                yield return  ThousandRowTestResultSet;
+                yield return  ZeroRowTestResultSet;
+                yield return  HundredKRowTestResultSet;
+                yield return  MillionRowTestResultSet;
```
Comment (MattIrv) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/QueryExecution/Common.cs]: minor: most of these have 2 spaces after `return`

Comment (ranasaria) [reply test/Microsoft.SqlTools.ServiceLayer.UnitTests/QueryExecution/Common.cs]: Fixed.

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs (hunk 1 above)
Comment (ranasaria) [reply src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs]: This method is callback delegate signature of a TimerCallback method which requires this parameter in the signature.

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs [hunk 3]
```diff
@@ -613,45 +617,70 @@ protected virtual void Dispose(bool disposing)
...
+        {
+            // Make the call to send current results and synchronously wait for them to finish
```
Comment (ranasaria) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs]: Wait for it to finish .. (not them)

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs [hunk 4]
```diff
@@ -613,45 +617,70 @@ protected virtual void Dispose(bool disposing)
...
                 //
-                await (ResultAvailable?.Invoke(currentResultSetSnapshot) ?? Task.CompletedTask);
-                ResultAvailable = null; // set this to null as we need to call ResultAvailable only once
+                ResultsTask =  (ResultAvailable?.Invoke(currentResultSetSnapshot) ?? Task.CompletedTask);
+                await ResultsTask;
+            }
+            else if (LastUpdatedSummary.Complete) // If last result summary sent had already set the Complete flag
+            {
+                // We do not need to do anything except that make sure that RowCount has not update since last send.
+                Debug.Assert(LastUpdatedSummary.RowCount == currentResultSetSnapshot.RowCount, $"Already reported rows should be equalt to current RowCount, if had alreaday send completion flag in last message, countReported:{LastUpdatedSummary.RowCount}, current total row count: {currentResultSetSnapshot.RowCount}, row count override: {currentResultSetSnapshot.rowCountOverride}, this.rowCountOverride: {this.rowCountOverride} and this.RowCount: {this.RowCount}, LastUpdatedSummary: {LastUpdatedSummary}");
```
Comment (ranasaria) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs]: typo: equalt 
typo: alreaday

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/QueryExecution/Common.cs (hunk 2 above)
Comment (ranasaria) [reply test/Microsoft.SqlTools.ServiceLayer.UnitTests/QueryExecution/Common.cs]: fixed.

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs (hunk 3 above)
Comment (ranasaria) [reply src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs]: fixed.

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs (hunk 4 above)
Comment (ranasaria) [reply src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs]: fixed

**Pull request merged**